src/workerd/io/trace-stream.c++

```diff
@@ -0,0 +1,821 @@
+  obj.set(js, "type"_kj, js.str("scheduled"_kj));
+  obj.set(js, "scheduledTime"_kj, js.date(info.scheduledTime));
+  obj.set(js, "cron"_kj, js.str(info.cron));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::AlarmEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("alarm"_kj));
+  obj.set(js, "scheduledTime"_kj, js.date(info.scheduledTime));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::QueueEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("queue"_kj));
+  obj.set(js, "queueName"_kj, js.str(info.queueName));
+  obj.set(js, "batchSize"_kj, js.num(info.batchSize));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::EmailEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("email"_kj));
+  obj.set(js, "mailFrom"_kj, js.str(info.mailFrom));
+  obj.set(js, "rcptTo"_kj, js.str(info.rcptTo));
+  obj.set(js, "rawSize"_kj, js.num(info.rawSize));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::TraceEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("trace"_kj));
+
+  auto names = KJ_MAP(trace, info.traces) -> jsg::JsValue {
+    KJ_IF_SOME(name, trace.scriptName) {
+      return js.str(name);
+    }
+    return js.null();
+  };
+  obj.set(js, "traces"_kj, js.arr(names));
+
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::HibernatableWebSocketEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("hibernatableWebSocket"_kj));
+
+  KJ_SWITCH_ONEOF(info.type) {
+    KJ_CASE_ONEOF(message, tracing::HibernatableWebSocketEventInfo::Message) {
+      auto mobj = js.obj();
+      mobj.set(js, "type"_kj, js.str("message"_kj));
+      obj.set(js, "info"_kj, mobj);
+    }
+    KJ_CASE_ONEOF(error, tracing::HibernatableWebSocketEventInfo::Error) {
+      auto mobj = js.obj();
+      mobj.set(js, "type"_kj, js.str("error"_kj));
+      obj.set(js, "info"_kj, mobj);
+    }
+    KJ_CASE_ONEOF(close, tracing::HibernatableWebSocketEventInfo::Close) {
+      auto mobj = js.obj();
+      mobj.set(js, "type"_kj, js.str("close"_kj));
+      mobj.set(js, "code"_kj, js.num(close.code));
+      mobj.set(js, "wasClean"_kj, js.boolean(close.wasClean));
+      obj.set(js, "info"_kj, mobj);
+    }
+  }
+
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::Resume& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("resume"_kj));
+
+  KJ_IF_SOME(attachment, info.attachment) {
+    jsg::Serializer::Released data{
+      .data = kj::heapArray<kj::byte>(attachment),
+    };
+    jsg::Deserializer deser(js, data);
+    obj.set(js, "attachment"_kj, deser.readValue(js));
+  }
+
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::CustomEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("custom"_kj));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const EventOutcome& outcome) {
```

Comment:
All values can be defined in constexpr evaluation and be put in a kj::Map or something to remove the repetition (and copying)

Reply:
The `js.str(...)` values cannot be made constexpr. Those are js heap allocated strings.